### PR TITLE
Add support for Spotify Authorization and Control

### PR DESCRIPTION
**Pull request by Behn4m**

### CODE REVIEW

File: components/HttpsCoponent_/HttpsRequest_.c [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+*@brief This function performs an HTTPS GET request to a specified server URL with the provided configuration.
+*@param[in] cfg The TLS configuration for the request.
+*@param[in] WEB_SERVER_URL The URL to which the request should be sent.
+*@param[in] REQUEST The HTTP request to be sent.
+*/
+static void https_get_request(esp_tls_cfg_t cfg, const char *WEB_SERVER_URL, const char *REQUEST)
+{
+    char buf[2500];
+    int ret, len;
```
Comment (Behn4m) [inline components/HttpsCoponent_/HttpsRequest_.c:79]: it's better to define variables in part of the code where you want to use them

File: components/HttpsCoponent_/HttpsRequest_.c [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    }
+    esp_tls_cfg_t cfg = {
+        .use_global_ca_store = true,
+    };
+    https_get_request(cfg, Web_URL, HttpsBuf);
+    esp_tls_free_global_ca_store();
+}
+
+#ifdef CONFIG_EXAMPLE_CLIENT_SESSION_TICKETS
+static void https_get_request_to_local_server(const char *url)
```
Comment (Behn4m) [inline components/HttpsCoponent_/HttpsRequest_.c:207]: is this function to "send request to local server" or "get request from local server"?
anyway, the name is too long,

File: components/HttpsCoponent_/HttpsRequest_.c
Comment (Behn4m) [inline components/HttpsCoponent_/HttpsRequest_.c:1]: some functions use lowerCamelCase and some use snake_case!

File: components/HttpsCoponent_/include/HttpsRequest_.h
Comment (Behn4m) [inline components/HttpsCoponent_/include/HttpsRequest_.h:1]: delete multiple new lines

File: components/SpotifyAuthorization/MakeSpotifyRequest.c [hunk 3]
```diff
@@ -0,0 +1,297 @@
+
+
```
Comment (Behn4m) [inline components/SpotifyAuthorization/MakeSpotifyRequest.c:2]: remove new lines here

File: components/SpotifyAuthorization/MakeSpotifyRequest.c [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+/**
+*@brief This function searches for specific patterns ('code' and 'state') within a character array and returns a boolean value indicating if either pattern was found.
+*@param[in] Res The character array to search within, and Res is response from first stage from spotify athurisiation
+*@param[in] SizeRes The size of the character array.
+*@return Returns true if either the 'code' or 'state' pattern was found, and false otherwise.
+*/
+bool FindCode(char *Res, uint16_t SizeRes)
+{
+
```
Comment (Behn4m) [inline components/SpotifyAuthorization/MakeSpotifyRequest.c:21]: remove the new line here

File: components/SpotifyAuthorization/MakeSpotifyRequest.c [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+            {
+                ESP_LOGI("Abbasi", "\twe find CODE !\n");
+                flg_findCode = 1;
+            }
+        }
+        if (Res[i] == 's')
+        {
+            if (Res[i + 1] == 't' && Res[i + 2] == 'a' && Res[i + 3] == 't' && Res[i + 3] == 'e')
+            {
+                ESP_LOGI("Abbasi", "\twe find State !\n");
```
Comment (Behn4m) [inline components/SpotifyAuthorization/MakeSpotifyRequest.c]: what is Abbasi? :D

Comment (Behn4m) [reply components/SpotifyAuthorization/MakeSpotifyRequest.c]: define a TAG, for example you can use Spotify

File: components/SpotifyAuthorization/MakeSpotifyRequest.c [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+/**
+
+*@brief This function sends a request to the Spotify API to perform a player command.
+*@param[in] Method_ The HTTP method to be used in the request (e.g., "POST", "PUT").
+*@param[in] Command_ The specific player command to be executed (e.g., "next", "previous", "play", "pause").
+*@param[out] Buf The character buffer to store the request and receive the response.
+*@param[in] SizeBuf The size of the character buffer.
+*@return This function does not return a value.
+*/
+void SendToPlayerCommand(const char *Method_, const char *Command_, char *Buf, size_t SizeBuf)
```
Comment (Behn4m) [inline components/SpotifyAuthorization/MakeSpotifyRequest.c]: unclear name, what does this function do? what is "Player" command?

File: components/SpotifyAuthorization/MakeSpotifyRequest.c [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+    MyHttpsHandler(Buf, SizeBuf, url, sizeof(url), server, sizeof(server));
+}
+/**
+
+*@brief This function sends a request to the Spotify API to skip to the next track in the player.
+*@return This function does not return a value.
+*/
+void SendRequestForNext()
+{
+    char Buf[2500];
```
Comment (Behn4m) [inline components/SpotifyAuthorization/MakeSpotifyRequest.c:160]: what is the "buf" used for? is it a dummy buffer or it stores the feedback?

File: components/SpotifyAuthorization/MakeSpotifyRequest.c [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+    SendToPlayerCommand("POST", "next", Buf, 2500);
+}
+/**
+
+*@brief This function sends a request to the Spotify API to skip to the previous track in the player.
+*@return This function does not return a value.
+*/
+void SendRequestForPrevious()
+{
+    char Buf[2500];
```
Comment (Behn4m) [inline components/SpotifyAuthorization/MakeSpotifyRequest.c:170]: same question on buf

File: components/SpotifyAuthorization/MakeSpotifyRequest.c [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+    SendToPlayerCommand("POST", "previous", Buf, 2500);
+}
+/**
+
+*@brief This function sends a request to the Spotify API to play the current track in the player.
+*@return This function does not return a value.
+*/
+void SendRequestForPlay()
+{
+    char Buf[2500];
```
Comment (Behn4m) [inline components/SpotifyAuthorization/MakeSpotifyRequest.c:180]: and again, what is buf for?

File: components/SpotifyAuthorization/MakeSpotifyRequest.c [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    SendToPlayerCommand("PUT", "pause", Buf, 2500);
+}
+/**
+
+*@brief This function sends a request to the Spotify API to retrieve the user's profile information.
+*@return This function does not return a value.
+*/
+void UserStatus()
```
Comment (Behn4m) [inline components/SpotifyAuthorization/MakeSpotifyRequest.c]: should be "GetUserStatus"

File: components/SpotifyAuthorization/MakeSpotifyRequest.c [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+/**
+ *@brief This function sends a request to the Spotify login API to exchange an authorization code for an access token.
+ *@param[in,out] Buf The character buffer to store the request and receive the response.
+ *@param[in] SizeBuf The size of the character buffer.
+ *@param[in] code is parameter that we give it before .
+ *@return This function does not return a value.
+ */
+void SendRequest_AndGiveTokenWithRefreshToken(char *Buf, size_t SizeBuf, char *RefreshToken_)
```
Comment (Behn4m) [inline components/SpotifyAuthorization/MakeSpotifyRequest.c:278]: this seems to be a good name for function. why you are not using this for all functions?

File: components/SpotifyAuthorization/SpotifyAuthorization.c [hunk 12]
```diff
@@ -0,0 +1,292 @@
+
```
Comment (Behn4m) [inline components/SpotifyAuthorization/SpotifyAuthorization.c:1]: remove new lines here

File: components/SpotifyAuthorization/SpotifyAuthorization.c [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+static const httpd_uri_t Responce_ = {
+    .uri = "/callback/",
+    .method = HTTP_GET,
+    .handler = HttpsUserCallBackFunc};
+/**
+ * @brief This function starts the web server for handling HTTPS requests.
+ *
+ * @return Returns the HTTP server handle if it is started successfully, or NULL otherwise.
+ */
+static httpd_handle_t StartWebServer(void)
```
Comment (Behn4m) [inline components/SpotifyAuthorization/SpotifyAuthorization.c:94]: Global variables should be defined at the beginning of the code

File: components/SpotifyAuthorization/SpotifyAuthorization.c [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+
+/**
+ * @brief This function is the handler for the connect event.
+ *
+ * @param[in] arg Pointer to the HTTP server handle.
+ * @param[in] event_base The event base.
+ * @param[in] event_id The event ID.
+ * @param[in] event_data The event data.
+ */
+static void connect_handler(void *arg, esp_event_base_t event_base,
```
Comment (Behn4m) [inline components/SpotifyAuthorization/SpotifyAuthorization.c]: what is "connect"?

File: components/SpotifyAuthorization/SpotifyAuthorization.c [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+}
+/**
+ * @brief This function is the handler for the disconnect event.
+ *
+ * @param[in] arg Pointer to the HTTP server handle.
+ * @param[in] event_base The event base.
+ * @param[in] event_id The event ID.
+ * @param[in] event_data The event data.
+ */
+static void disconnect_handler(void *arg, esp_event_base_t event_base,
```
Comment (Behn4m) [inline components/SpotifyAuthorization/SpotifyAuthorization.c]: what is "disconnect"?

File: components/SpotifyAuthorization/SpotifyAuthorization.c [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+        ESP_LOGI(TAG_APP, "MDNS Init failed: %d\n", err);
+        return;
+    }
+    mdns_hostname_set("deskhub");
+    mdns_instance_name_set("Behnam's ESP32 Thing");
+}
+/**
+ * @brief This function handles the Spotify authorization process.
+ */
+void SpotifyComponent()
```
Comment (Behn4m) [inline components/SpotifyAuthorization/SpotifyAuthorization.c:173]: the function name should be something like "SpotifyAuthorization"

File: components/SpotifyAuthorization/SpotifyAuthorization.c [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+    FindCodeSemaphore = xSemaphoreCreateBinary();
+    FailToFindCodeSemaphore = xSemaphoreCreateBinary();
+    xTaskCreate(&HttpServer, "HttpServer", 10000, NULL, 1, NULL);
+}
+
+/**
+ * @brief This function is the entry point for handling HTTPS requests for Spotify authorization.
+ * @param[in] pvparameters we need it because its Task !
+ */
+static void HttpServer(void *pvparameters)
```
Comment (Behn4m) [inline components/SpotifyAuthorization/SpotifyAuthorization.c]: name is not clear to me

File: components/WiFiComponent/include/WiFiConfig_.h [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+
+
+void wifi_connection();
+
+
+
+
+
+
+
```
Comment (Behn4m) [inline components/WiFiComponent/include/WiFiConfig_.h:16]: lots of new lines!!!

File: main/GlobalInit.c [hunk 19]
```diff
@@ -0,0 +1,20 @@
+
+#include"main.h"
+
+extern QueueHandle_t BufQueue1;
+extern SemaphoreHandle_t GetResponseSemaphore;
+
+
```
Comment (Behn4m) [inline main/GlobalInit.c:7]: lots of new lines!!!

File: main/main.c [hunk 20]
```diff
@@ -1,13 +1,24 @@
...
+
+
+// wifi ssid is "Hardware10"  and wifi pass is "87654321"
+
+
 void app_main(void)
-{   
-    nvsFlashInit();    
+{
+    MyInit();
```
Comment (Behn4m) [inline main/main.c]: "MyInit" ????

**Review by Behn4m — COMMENTED** [review]
Lots of comments, please check carefully :)

File: components/SpotifyAuthorization/SpotifyAuthorization.c (hunk 16 above)
Comment (MohammadAbbasi76) [reply components/SpotifyAuthorization/SpotifyAuthorization.c:173]: I will change it

File: components/SpotifyAuthorization/SpotifyAuthorization.c (hunk 17 above)
Comment (MohammadAbbasi76) [reply components/SpotifyAuthorization/SpotifyAuthorization.c]: I will change it to  "HttpLocaleServer"

File: components/HttpsCoponent_/HttpsRequest_.c (hunk 1 above)
Comment (MohammadAbbasi76) [reply components/HttpsCoponent_/HttpsRequest_.c:79]: this is esp idf s function !

File: components/HttpsCoponent_/HttpsRequest_.c (hunk 2 above)
Comment (MohammadAbbasi76) [reply components/HttpsCoponent_/HttpsRequest_.c:207]: if I add some comment that show what function exactly do is better , I will do it

File: components/HttpsCoponent_/HttpsRequest_.c
Comment (MohammadAbbasi76) [reply components/HttpsCoponent_/HttpsRequest_.c:1]: because I write them and this is my way for naming

Comment (MohammadAbbasi76) [reply components/HttpsCoponent_/HttpsRequest_.c:1]: my way was Camelcase

File: components/HttpsCoponent_/HttpsRequest_.c (hunk 2 above)
Comment (Behn4m) [reply components/HttpsCoponent_/HttpsRequest_.c:207]: the name needs to be clear also

File: components/HttpsCoponent_/HttpsRequest_.c
Comment (Behn4m) [reply components/HttpsCoponent_/HttpsRequest_.c:1]: so, if they are IDF functions, why you define them again?

File: components/HttpsCoponent_/HttpsRequest_.c (hunk 1 above)
Comment (Behn4m) [reply components/HttpsCoponent_/HttpsRequest_.c:79]: so, why you are defining it? is it an overwrite?

**Pull request closed without merging**